java/ext/javadoctest.py

```diff
@@ -0,0 +1,57 @@
+import subprocess
+import os
+from sphinx.ext.doctest import TestcodeDirective, TestoutputDirective, DocTestBuilder, Any, Dict, doctest, sphinx
+from sphinx.locale import __
+from subprocess import Popen, PIPE
+
+class JavaDocTestBuilder(DocTestBuilder):
+    """
+    Runs java test snippets in the documentation.
+    """
+    name = 'javadoctest'
+    epilog = __('Java testing of doctests in the sources finished, look at the '
+                'results in %(outdir)s/output.txt.')
+    def compile(self, code: str, name: str, type: str, flags: Any, dont_inherit: bool) -> Any:
+        # go to project that contains all your arrow maven dependencies
+        dir_arrow_maven_project = os.path.dirname(os.getcwd())+'/java/source/demo'
+
+        # create list of all arrow jar dependencies
+        proc_mvn_dependency = subprocess.Popen(['mvn', '-q', 'dependency:build-classpath', '-DincludeTypes=jar', '-Dmdep.outputFile=.cp.tmp'], cwd=dir_arrow_maven_project, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
+        stdout_mvn, stderr_mvn = proc_mvn_dependency.communicate()
+        if not os.path.exists(dir_arrow_maven_project + '/.cp.tmp'):
+            raise RuntimeError(__('invalid process to create jshell dependencies library'))
+
+        # get list of all arrow jar dependencies
+        proc_get_dependency = subprocess.Popen(['cat', '.cp.tmp'], cwd=dir_arrow_maven_project, stdout=subprocess.PIPE, text=True)
+        stdout_dependency, stderr_dependency = proc_get_dependency.communicate()
+        if not stdout_dependency:
+            raise RuntimeError(__('invalid process to list jshell dependencies library'))
+
+        # execute java testing code thru jshell and read output
+        proc_java_arrow_code = subprocess.Popen(['echo', '' + code], stdout=subprocess.PIPE)
+        proc_jshell_process = subprocess.Popen(['jshell', '--class-path', stdout_dependency, '-'], stdin=proc_java_arrow_code.stdout, stdout=subprocess.PIPE, text=True)
```

Comment:
Is Jshell really unable to take in a file?

Regardless, there's no need to use echo. `subprocess` lets you pass in `subprocess.PIPE` to `stdin` and then we can write to it directly.

Reply:
> Is Jshell really unable to take in a file?

Jshell is able to execute a file thru: jshell filename.jsh and it execute all the java code inside filename.jsh 

But in our case we are not offering a filename.jsh file per our recipe, we are putting our code inside .rst files

> 
> Regardless, there's no need to use echo. `subprocess` lets you pass in `subprocess.PIPE` to `stdin` and then we can write to it directly.

Let me test that

Reply:
I just try to delete echo, but this is not working as expected.